File: Algorithm/Recursion & Backtracking/N-Queens.py
board=[[False]*8 for _ in range(8)]

def printBoard():
  for i in range(8):
    for j in range(8):
      print("Q" if board[i][j] else ".",end="")
    print()
  
  print("-"*30)


def safe(row ,col):
  
  for i in range(8):
    if board[row][i]:
      return False
  
  for i in range(8):
    if board[i][col]:
      return False
  
  i, j = row, col
  while i >= 0 and j >= 0:
      if board[i][j]:
          return False
      i -= 1
      j -= 1
    
  i, j = row, col
  while i < 8 and j < 8:
      if board[i][j]:
          return False
      i += 1
      j += 1
    
  i, j = row, col
  while i >= 0 and j < 8:
      if board[i][j]:
          return False
      i -= 1
      j += 1
  
  i, j = row, col
  while i < 8 and j >= 0:
      if board[i][j]:
          return False
      i += 1
      j -= 1
  
  return True

def N_Queens(row):
  if row==8 :
    printBoard()
    return 
  
  for col in range(8):
    if safe(row,col):
      board[row][col]=True
      
      N_Queens(row+1)
      
      board[row][col]=False

# Main

N_Queens(0)

File: Algorithm/Recursion & Backtracking/N-Queens.cpp
#include <iostream>
using namespace std;
bool board[8][8];

void printBoard() {
    for(int row=0;row<8;row++){
        for(int col=0;col<8;col++){
            cout << (board[row][col] ? "Q " : ". ");
        }
        cout << endl;
    }
    cout << "---------------\n";
}

bool safe(int row,int col){
  
  for(int j=0;j<8;j++){
        if(board[row][j])
        return false;
    }
  
  for(int j=0;j<8;j++){
       if(board[j][col])
       return false;
    }
    
  
  for(int i=row,j=col;i>=0 && j>=0;i--,j--){
    if(board[i][j])
    return false;
  }
   
  for(int i=row,j=col;i<8 && j<8;i++,j++){
    if(board[i][j])
    return false;
  }
  
  for(int i=row,j=col;i>=0 && j<8;i--,j++){
    if(board[i][j])
    return false;
  }
  
  
  for(int i=row,j=col;i<8 && j>=0;i++,j--){
    if(board[i][j])
    return false;
  }
  
  return true;
  
}
void N_Queens(int row){
  
  if(row==8){
    printBoard();
    return ;
  }
  
  for(int col=0;col<8;col++){
    if(safe(row,col)){
      board[row][col]=true;
      N_Queens(row+1);
      board[row][col]=false;
    }
    
  }
  
}
int main() {
  N_Queens(0);
    return 0;
}